src/sync.rs

```diff
@@ -473,6 +480,58 @@ where
     }
 }
 
+/// Helper trait for `Once` to help avoid deadlocking when using a `Once` when attached to a
+/// Python thread.
+pub trait OnceExt: Sealed {
+    /// Similar to [`call_once`][Once::call_once], but releases the Python GIL temporarily
+    /// if blocking on another thread currently calling this `Once`.
+    fn call_once_py_attached(&self, py: Python<'_>, f: impl FnOnce());
+
+    /// Similar to [`call_once_force`][Once::call_once_force], but releases the Python GIL
+    /// temporarily if blocking on another thread currently calling this `Once`.
+    fn call_once_force_py_attached(&self, py: Python<'_>, f: impl FnOnce(&OnceState));
+}
+
+impl OnceExt for Once {
+    fn call_once_py_attached(&self, _py: Python<'_>, f: impl FnOnce()) {
+        if self.is_completed() {
+            return;
+        }
+
+        // Safety: we are currently attached to the GIL, and we expect to block. We will save
+        // the current thread state and restore it as soon as we are done blocking.
+        let mut ts = Some(unsafe { ffi::PyEval_SaveThread() });
+
+        self.call_once(|| {
+            unsafe { ffi::PyEval_RestoreThread(ts.take().unwrap()) };
+            f();
+        });
+        if let Some(ts) = ts {
+            // Some other thread filled this Once, so we need to restore the GIL state.
+            unsafe { ffi::PyEval_RestoreThread(ts) };
+        }
+    }
```

Comment:
I wrote a test for this trait based on the standard library example for `try_once_force` and it crashed Python because the thread state was corrupted. I think this code won't restore the thread state correctly if a panic happens inside `f()` and the fix is to use the RAII guard pattern. Assuming that's right, I'll go ahead and push a fix along with the test.

Reply:
Does the guard fix your problem? That'd be surprising to me, because I reviewed the previous implementation and believed it was sound.

Reply:
Oh nevermind; I read the Once documentation and if it is poisoned it will panic before the closure is called.